pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -727,7 +728,7 @@ func (c *Reconciler) createTaskRun(ctx context.Context, rprt *resources.Resolved
 		// We pass the entire, original task ref because it may contain additional references like a Bundle url.
 		tr.Spec.TaskRef = rprt.PipelineTask.TaskRef
 	} else if rprt.ResolvedTaskResources.TaskSpec != nil {
-		tr.Spec.TaskSpec = rprt.ResolvedTaskResources.TaskSpec
+		tr.Spec.TaskSpec = tresources.ApplyPipelineTaskContexts(rprt.ResolvedTaskResources.TaskSpec, rprt.PipelineTask)
```

Comment:
Ok, but I notice some variables like $(context.taskRun.uid) can be directly used in the taskSpec. There is an example to do so [examples/v1beta1/taskruns/using_context_variables.yaml](https://github.com/tektoncd/pipeline/blob/main/examples/v1beta1/taskruns/using_context_variables.yaml)

Reply:
You're absolutely right about `TaskRuns` with embedded `taskSpecs`.

Here we deal with `Pipeline` -> embedded `PipelineTask` -> embedded `taskSpec`. No other pipeline context variable in this situation gets replaced. Let's look at another example:

```yaml
kind: Pipeline
apiVersion: tekton.dev/v1beta1
metadata:
  name: foo
spec:
  tasks:
  - name: t
    taskSpec:
      steps:
      - image: alpine
        script: |
          echo $(context.pipelineRun.uid)
---
kind: PipelineRun
apiVersion: tekton.dev/v1beta1
metadata:
  name: foo-run
spec:
  pipelineRef:
    name: foo
```

Trying to access `$(context.pipelineRun.uid)` directly in this way doesn't work. There's an error because the variable did not get replaced in the `taskSpec`:

```
[t : unnamed-0] + context.pipelineRun.uid
[t : unnamed-0] /tekton/scripts/script-0-8tpb7: line 3: context.pipelineRun.uid: not found
```

I do think we should improve this - the pipeline context variables should be usable in the `taskSpec`.  But I think that should happen for all the context variables at once, so possibly in another PR to discuss?